Ensure only primary sender drives slot ownership updates

PingXie opened the pull request:
Fixes a regression introduced in PR #445, which allowed a message from a replica
to update the slot ownership of its primary. The regression results in a
`replicaof` cycle, causing server crashes due to the cycle detection assert. The
fix restores the previous behavior where only primary senders can trigger
`clusterUpdateSlotsConfigWith`.

Additional changes:

* Handling of primaries without slots is obsoleted by new handling of when a
  sender that was a replica announces that it is now a primary.
* Replication loop detection code is unchanged but shifted downwards.
* Some variables are renamed for better readability and some are introduced to
  avoid repeated memcmp() calls.

Fixes #753.

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -2870,6 +2870,9 @@ int clusterProcessPacket(clusterLink *link) {
     uint16_t flags = ntohs(hdr->flags);
     uint64_t senderCurrentEpoch = 0, senderConfigEpoch = 0;
     clusterNode *sender = getNodeFromLinkAndMsg(link, hdr);
+    int sender_claims_primary = !memcmp(hdr->replicaof, CLUSTER_NODE_NULL_NAME, CLUSTER_NAMELEN);
+    int sender_was_replica = sender && nodeIsReplica(sender);
+    int sender_was_primary = sender && nodeIsPrimary(sender);
```
Good to introduce these variables!

I'm not sure about the name `sender_claims_primary`. It sounds a bit like it's claiming another node as its primary.

How about `sender_is_primary`? We always trust the sender, right?

Otherwise something like `sender_claims_to_be_primary`(?), but it's a bit long.

PingXie replied:
I thought `sender_claims_primary` is `sender_claims_to_be_primary` but sure I can go with the longer name too, especially when cluster states are involved.

zuiderkwast replied:
Yes, I think `sender_claims_to_be_primary` is easier and faster to understand.

Most often, the word claim is used when you claim some object, like a slot. To claim that you are something is not the most common use of the word claim. With "claims to be", it's immediately very clear to me. I'd understand the variable name in 1 seconds instead of 5.

PingXie replied:
sounds good. will update to `sender_claims_to_be_primary`

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -3051,19 +3053,17 @@ int clusterProcessPacket(clusterLink *link) {
             } else {
                 /* Node is a replica. */
-                clusterNode *primary = clusterLookupNode(hdr->replicaof, CLUSTER_NAMELEN);
+                clusterNode *new_primary = clusterLookupNode(hdr->replicaof, CLUSTER_NAMELEN);
```
In most cases, this is not a *new* primary, but just the primary of the sender. "New" indicates some change.

I suggest keeping the name `primary` for this variable.

PingXie replied:
I am trying to make a distinction of what "I"/this node sees as a primary of the sender, which can be stale, and what the sender claims its primary is. 

`primary` is way overloaded in this function so I wouldn't want to go back to it but how about `sender_claimed_primary`?

zuiderkwast replied:
I think a confusion here is that we use "sender" for two things:

1. The stuff that we find in the cluster bus msg, i.e. `hrd`.
2. The node that we already have stored and looked up as `clusterLookupNode(hdr->sender, CLUSTER_NAMELEN)`

What if we can use two separate words for these?

PingXie replied:
>I think a confusion here is that we use "sender" for two things:

Spot on :-). This is one of the things that bugs me all the time.

>What if we can use two separate words for these?

Agreed. I don't claim to have the best names but what do you think about?

1. `sender_message` 
2. `sender_node`

PingXie replied:
I have a different proposal. Let me update the code and it will be easier to review them in the context.

hpatro commented on src/cluster_legacy.c:
```diff
@@ -2887,8 +2890,7 @@ int clusterProcessPacket(clusterLink *link) {
         /* Update the sender configEpoch if it is a primary publishing a newer one. */
-        if (!memcmp(hdr->replicaof, CLUSTER_NODE_NULL_NAME, sizeof(hdr->replicaof)) &&
-            senderConfigEpoch > sender->configEpoch) {
+        if (sender_claims_primary && senderConfigEpoch > sender->configEpoch) {
```
Should we move this entire condition to a variable and name it as `sender_is_primary`?

```
int sender_is_primary = (!memcmp(hdr->replicaof, CLUSTER_NODE_NULL_NAME, CLUSTER_NAMELEN) && senderConfigEpoch > sender->configEpoch);
```

PingXie replied:
`sender_is_new_primary` sounds like a better name to me. Let me play with it a bit next.

zuiderkwast replied:
It can be a primary even if `senderConfigEpoch == sender->configEpoch`, can't it?

Only if it was a replica before and a primary now, we can require that the config epoch is bumped.

PingXie replied:
>It can be a primary even if senderConfigEpoch == sender->configEpoch, can't it?

Not sure I follow. The "primary" part is inferred from `memcmp(hdr->replicaof, CLUSTER_NODE_NULL_NAME, CLUSTER_NAMELEN)`. And then the epoch check is separate. The idea is to make sure that we are not looking at a previous election result.

>Only if it was a replica before and a primary now, we can require that the config epoch is bumped.

Yes there needs to be a epoch bump for a "promotion". Epoch bump is part of the election protocol than an add-on requirement. 

>Should we move this entire condition to a variable and name it as sender_is_primary?

The new name `sender_claims_to_be_primary` is a lot clearer now and also very long. I haven't been able to come up with a better name that would allow me to encode the "more recent" property, which would probably read like `sender_cliams_to_be_a_more_recent_primary_from_my_perspective` :D. I think I will keep `sender_claims_to_be_primary` for now.

zuiderkwast replied:
> > It can be a primary even if senderConfigEpoch == sender->configEpoch, can't it?
> 
> Not sure I follow.

This was in response to @hpatro's suggestion above, which includes `senderConfigEpoch > sender->configEpoch`:

    int sender_is_primary = (!memcmp(hdr->replicaof, CLUSTER_NODE_NULL_NAME, CLUSTER_NAMELEN) && senderConfigEpoch > sender->configEpoch);

hpatro commented on src/cluster_legacy.c:
```diff
@@ -3128,63 +3128,38 @@ int clusterProcessPacket(clusterLink *link) {
+         * for it or if there was a failover in the sender's shard. Check
+         * this ASAP to avoid other computational expensive checks later.*/
+
+        if (sender && sender_claims_primary &&
```
`sender_claims_primary` in itself is not good enough if it's epoch isn't the latest, right?

PingXie replied:
Yes but the epoch check is done in `clusterUpdateSlotsConfigWith` today. I need to think a bit whether we can safely pull it out of `clusterUpdateSlotsConfigWith`.

hpatro commented on src/cluster_legacy.c:
```diff
@@ -3203,10 +3178,21 @@ int clusterProcessPacket(clusterLink *link) {
 
+        /* Explicitly check for a replication loop before attempting the replication
+         * chain folding logic. */
+        if (myself->replicaof && myself->replicaof->replicaof && myself->replicaof->replicaof != myself) {
```
Ah, this is becoming so complex.


Where do we handle the cyclic scenario?
```
myself->replicaof && myself->replicaof->replicaof && myself->replicaof->replicaof == myself
```

PingXie replied:
This is not for the cyclic replication handling. This is the replication chain reduction/compression logic. 

We handle/prevent the cyclic by allowing primaries only to enter `clusterUpdateSlotsConfigWith`

PingXie replied:
BTW, this is not new logic. It is shifted downwards.

hpatro reviewed (commented):
Question from the issue (repeating here):

 Is it any problematic to handle topology update from replica with higher config epoch?

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -3128,63 +3128,38 @@ int clusterProcessPacket(clusterLink *link) {
-            if (sender_primary) {
-                dirty_slots = memcmp(sender_primary->slots, hdr->myslots, sizeof(hdr->myslots)) != 0;
-
-                /* Force dirty when the sending shard owns no slots so that
-                 * we have a chance to examine and repair slot migrating/importing
-                 * states that involve empty shards. */
-                dirty_slots |= sender_primary->numslots == 0;
```
Why is this deleted?

PingXie replied:
This is a new behavior introduced in #445. It is not generic enough to work in slot migration scenarios because it only covers the empty shard case. I generalized it to `sender_was_replica`.

zuiderkwast commented on src/cluster_legacy.c:
```diff
@@ -3128,63 +3128,38 @@ int clusterProcessPacket(clusterLink *link) {
-        clusterNode *sender_primary = NULL; /* Sender or its primary if replica. */
-        int dirty_slots = 0;                /* Sender claimed slots don't match my view? */
-
-        if (sender) {
-            sender_primary = clusterNodeIsPrimary(sender) ? sender : sender->replicaof;
-            if (sender_primary) {
-                dirty_slots = memcmp(sender_primary->slots, hdr->myslots, sizeof(hdr->myslots)) != 0;
```
So if a node is a replica, it still reports its slots and we used this to update the slots for its primary? And we don't want to do that anymore?

PingXie replied:
>So if a node is a replica, it still reports its slots and we used this to update the slots for its primary? 

No - this was the regression I introduced in #445. 

>And we don't want to do that anymore?

Yup

zuiderkwast reviewed (commented):
The change seems to make sense but I don't fully understand the fix and how it solves the problem. The essence of the problem can be seen in the diff screenshot in #753? Please update the PR description to independently describe the change, rather than only referring to the issue.

It's harder to review the fix when there's refactoring in the same commit. It's easier if the refacoring and the acutal fix are at least in two separate commits.

(I imagine Oran Agra would have rejected the renaming of variables, to avoid any kind of unnecessary changes.)

PingXie commented:
> Is it any problematic to handle topology update from replica with higher config epoch?

Yeah I think it will work but I feel that it might involve a bigger change. This change on the other hand is more scoped, relatively speaking. We can definitely explore this idea separately/incrementally.

PingXie commented:
> The change seems to make sense but I don't fully understand the fix and how it solves the problem. The essence of the problem can be seen in the diff screenshot in #753? Please update the PR description to independently describe the change, rather than only referring to the issue.

Yes - replicas updating the slot ownership on behalf of their primaries is the regression (or behavior change) introduced by #445.

> It's harder to review the fix when there's refactoring in the same commit. It's easier if the refacoring and the acutal fix are at least in two separate commits.
> 
> (I imagine Oran Agra would have rejected the renaming of variables, to avoid any kind of unnecessary changes.)

I don't agree with the "unnecessary changes". All these changes are either a continuation/fine-tune of #445 or for a good reason of improving readability. I have gone through this code path more times than I can count but I still get lost easily because the naming is just so confusing. A blank statement of "unnecessary changes" is just saying "I don't know what I am doing" - sorry for the unnecessarily strong response :)

zuiderkwast commented:
> > It's harder to review the fix when there's refactoring in the same commit. It's easier if the refacoring and the acutal fix are at least in two separate commits.
> > (I imagine Oran Agra would have rejected the renaming of variables, to avoid any kind of unnecessary changes.)
> 
> I don't agree with the "unnecessary changes". All these changes are either a continuation/fine-tune of #445 or for a good reason of improving readability. I have gone through this code path more times than I can count but I still get lost easily because the naming is just so confusing.

Sorry for causing frustration! I know you've very familiar with this code. More so than the rest of us.

With "unnecessary changes", I just meant changes that don't change the logic, i.e. refactoring, not strictly necessary to fix the bug. I didn't say I'm against it. Only that I'd prefer them in separate commits, since it would have made it easier for me to read the diff and spot the actual behaviour change. The diff is +68 −82, while the actual fix is ~5 lines.

Even the title "ensure only primary sender drives..." seemed a bit cryptic to me at first, but of course it's obvious to anyone who has this code fresh in memory. (My though processes involved "What's a primary sender again... Hm.. right, it's when we're receiving a packet on the cluster bus from a primary.")

I'm trying to make reviewing easier for myself by requiring more from the contributors. Not sure if it's sane but I want to be able to handle more PRs faster without being burnt out. I should know you're in the same position.

PingXie commented:
> Sorry for causing frustration! I know you've very familiar with this code. More so than the rest of us.

No worries - we are doing all our job.

> With "unnecessary changes", I just meant changes that don't change the logic, i.e. refactoring, not strictly necessary to fix the bug. I didn't say I'm against it. Only that I'd prefer them in separate commits, since it would have made it easier for me to read the diff and spot the actual behaviour change. The diff is +68 −82, while the actual fix is ~5 lines.

I disagree. I don't consider this a refactoring, which I agree is better handled in a separate PR. There is a reason why I introduced the regression in the first place. Our discussions on the names show the exact problem of this code not having established clear mental concepts and this was partly the reason why I introduced the regression in the first place. Making these changes in a separate PR loses the exact context of why these changes are needed. And refactoring PRs don't normally get the right amount of scrutinization, which is what I want for this PR. I am all for any nitpick that could help improve the code quality in such a critical part of the code base.

> 
> Even the title "ensure only primary sender drives..." seemed a bit cryptic to me at first, but of course it's obvious to anyone who has this code fresh in memory. (My though processes involved "What's a primary sender again... Hm.. right, it's when we're receiving a packet on the cluster bus from a primary.")
> 
> I'm trying to make reviewing easier for myself by requiring more from the contributors. Not sure if it's sane but I want to be able to handle more PRs faster without being burnt out. I should know you're in the same position.

I believe I have explained the issue to my best ability in #753, which I linked to this PR as well. Have you got a chance to check it out? I have a tendency to split the bug and PR. Will make sure to carry the context over for this (and future) PR.

PingXie commented:
I have updated the names per recommendation. I think they look clearer now. @zuiderkwast PTAL. 

Also I think there is another (0-day) bug. We never check the sender's config epoch before accepting its claim of being a primary. I will address it in a separate PR :).

https://github.com/valkey-io/valkey/blob/unstable/src/cluster_legacy.c#L3119

zuiderkwast reviewed (approved):
Great! These "sender-claimed" names definitely make things more clear.

> Also I think there is another (0-day) bug. We never check the sender's config epoch before accepting its claim of being a primary. I will address it in a separate PR :).

OK. I wouldn't mind having it in the same PR (just a separate commit within the PR for easier reading). But a separate PR is also good to if we want to backport it or something.

> I don't consider this a refactoring, which I agree is better handled in a separate PR.

I never said separate PR. That would be excessive. I just meant separate commits within the same PR. It might have made it easier to review, though this may just have been me trying to find excuses for not understanding the fix.

> I believe I have explained the issue to my best ability in #753, which I linked to this PR as well. Have you got a chance to check it out? I have a tendency to split the bug and PR. Will make sure to carry the context over for this (and future) PR.

Yes, the issue is very good. The problem description and the description of the change are not the same things though. In the PR I think it's good to describe the actual change. The problem description doesn't need to be repeated in detail in the PR when it's explained in the issue.

For the PR description, I'd be happy just what you wrote under "Fix:", and possibly mentioning the additional changes to make it easier to understand the diff. No need to update it now, but I'm copy-pasting some sentences here just to show what I mean, what I think is enough, yet useful for reviewing and also enough to have in the commit log after merging:

```
Fixes a regression introduced in PR #445, which allowed a message from a replica
to update the slot ownership of its primary. The regression results in a
replicaof cycle, causing server crashes due to the cycle detection assert. The
fix restores the previous behavior where only primary senders can trigger
clusterUpdateSlotsConfigWith.

Additional changes:

* Handling of primaries without slots is obsoleted by new handling of when a
  sender that was a replica announces that it is now a primary.
* Replication loop detection code is unchanged but shifted downwards.
* Some variables are renamed for better readability and some are introduced to
  avoid repeated memcmp() calls.
```

PingXie commented:
Thanks @zuiderkwast! 


> OK. I wouldn't mind having it in the same PR (just a separate commit within the PR for easier reading). But a separate PR is also good to if we want to backport it or something.

Let me start a new PR on this one. It is probably easier that way.

> I never said separate PR. That would be excessive. I just meant separate commits within the same PR. It might have made it easier to review, though this may just have been me trying to find excuses for not understanding the fix.

Got it. My bad. I misunderstood. Reviewing commits is a good point.

> The problem description and the description of the change are not the same things though. In the PR I think it's good to describe the actual change. The problem description doesn't need to be repeated in detail in the PR when it's explained in the issue.

> For the PR description, I'd be happy just what you wrote under "Fix:", and possibly mentioning the additional changes to make it easier to understand the diff. No need to update it now, but I'm copy-pasting some sentences here just to show what I mean, what I think is enough, yet useful for reviewing and also enough to have in the commit log after merging

Make sense. I can do that too.

madolson commented on src/cluster_legacy.c:
```diff
@@ -2998,13 +3001,13 @@ int clusterProcessPacket(clusterLink *link) {
-            getClientPortFromClusterMsg(hdr, &node->tls_port, &node->tcp_port);
-            node->cport = ntohs(hdr->cport);
-            clusterAddNode(node);
+            new_sender_node = createClusterNode(NULL, CLUSTER_NODE_HANDSHAKE);
+            serverAssert(nodeIp2String(new_sender_node->ip, link, hdr->myip) == C_OK);
+            getClientPortFromClusterMsg(hdr, &new_sender_node->tls_port, &new_sender_node->tcp_port);
+            new_sender_node->cport = ntohs(hdr->cport);
+            clusterAddNode(new_sender_node);
```
I think these changes are not useful. "New sender node" adds more semantic meaning than before and I'm not sure what it means.

PingXie replied:
I am trying to convey the nature of this variable as a temp/local one. `node` to me sounds like a name that we will keep for the bigger scope. Will rename it to `n`. Thoughts?

madolson replied:
> node to me sounds like a name that we will keep for the bigger scope.

I would say n < new_sender_node < node. I was arguing before that this felt like a completely unrelated change. My general stance is I think we can improve the wording of stuff we are changing, but we shouldn't change unrelated stuff in the PR?

madolson commented on src/cluster_legacy.c:
```diff
@@ -2930,8 +2930,11 @@ int clusterProcessPacket(clusterLink *link) {
+    uint64_t sender_claimed_current_epoch = 0, sender_claimed_config_epoch = 0;
     clusterNode *sender = getNodeFromLinkAndMsg(link, hdr);
+    int sender_claims_to_be_primary = !memcmp(hdr->replicaof, CLUSTER_NODE_NULL_NAME, CLUSTER_NAMELEN);
+    int sender_was_replica = sender && nodeIsReplica(sender);
```
`sender_was_replica` to me implies it no longer is a replica, which is not true. `sender_last_reported_as_replica` feels verbose though.

PingXie replied:
I like `sender_last_reported_as_replica`. I don't mind it being verbose in this context.

madolson commented:
> OK. I wouldn't mind having it in the same PR (just a separate commit within the PR for easier reading). But a separate PR is also good to if we want to backport it or something.

I want it in a separate PR. I'm going to fairly strongly disagree with Viktor that I think in this specific case the refactor made the change harder for me to follow. I think that is because I'm more familiar with the code, outside of the newly introduced variables I found everything harder to follow.

madolson commented:
> Yes - replicas updating the slot ownership on behalf of their primaries is the regression (or behavior change) introduced by https://github.com/valkey-io/valkey/pull/445.

Yeah, allowing replicas to update slot ownership on behalf of their primaries is still the change that makes me feel uncomfortable.

PingXie commented:
> I think that is because I'm more familiar with the code, outside of the newly introduced variables I found everything harder to follow

This is likely proximity bias :). I would suggest also reviewing the function **without** diff'ing line by line, and then back the diff. This is how I come to being at peace with it. Let me know what you think afterwards.

madolson commented:
> This is likely proximity bias

Everyone is biased.

madolson commented:
> Let me know what you think afterwards.

I'll stand behind my original statement that the refactoring made evaluating the technical changes more difficult, and would still prefer it in a separate PR. I also think that is a better broad decision to take as a team. The problem with a separate commit is that we squash and merge, and lose the commit. 

In some cases you are introducing new concepts, and for that the new names make sense. But in other it just seems like you are renaming variables.

madolson reviewed (commented):
I think the change makes sense. We probably should avoid taking input from the replica about three state of its primaries as much as possible.

hpatro reviewed (approved):
Mostly LGTM. 

@PingXie @madolson Could we converge and close the PR out? I will further merge unstable to #573 and get it ready.

PingXie commented:
> I'll stand behind my original statement that the refactoring made evaluating the technical changes more difficult, and would still prefer it in a separate PR. I also think that is a better broad decision to take as a team. The problem with a separate commit is that we squash and merge, and lose the commit.

Sure. I can do a separate PR in the future.

>Could we converge and close the PR out?

Will clean this PR up in a bit.

hpatro commented on src/cluster_legacy.c:
```diff
@@ -2995,13 +2998,13 @@ int clusterProcessPacket(clusterLink *link) {
-            node = createClusterNode(NULL, CLUSTER_NODE_HANDSHAKE);
-            serverAssert(nodeIp2String(node->ip, link, hdr->myip) == C_OK);
-            getClientPortFromClusterMsg(hdr, &node->tls_port, &node->tcp_port);
-            node->cport = ntohs(hdr->cport);
-            clusterAddNode(node);
+            n = createClusterNode(NULL, CLUSTER_NODE_HANDSHAKE);
+            serverAssert(nodeIp2String(n->ip, link, hdr->myip) == C_OK);
+            getClientPortFromClusterMsg(hdr, &n->tls_port, &n->tcp_port);
+            n->cport = ntohs(hdr->cport);
+            clusterAddNode(n);
```
Do we need to update these?

PingXie replied:
strictly speaking no and so are the remaining name changes. The point that I am trying to get across is "does it help?". If folks believe readability gets worse, I will revert. Otherwise, I just don't see when we get to improve the readability. It is not like we would dedicate PRs just for improving readability :-)

hpatro replied:
Would have preferred `node` while writing from scratch. Working around Redis/Valkey has made me not modify lines not related to my PR.

Would let the approvers decide. I would prefer merging this out :)

madolson replied:
> It is not like we would dedicate PRs just for improving readability :-)

I would merge a PR that just makes readability better, I also think you would too, so I don't think this is true?

PingXie replied:
>so I don't think this is true?

Yes I do think it helps and we have used short names for temps pretty extensively in the code base, especially `n`. that said, I don't want to block the more important work on such a trivial thing. I reverted it back. Let's merge and move on?
